include/irr/asset/CTexturePacker.h

```diff
@@ -216,53 +206,79 @@ class ICPUTexturePacker : protected ITexturePacker
-            }
-        }
-        //pack tail mips into 1 page (if needed)
-        if (offsets->back().hasMipTailAddr())
-        {
-            const core::vector3du32_SIMD texOffset = pageCoords((*offsets).back().getMipTailAddr(), m_pgSzxy);
-            //TODO
-        }
-        //fill page table
-        const VkExtent3D extent = _img->getCreationParameters().extent;
-        for (uint32_t i = 0u; i < offsets->size(); ++i)
+        const uint32_t texelSz = getTexelOrBlockBytesize(m_physAddrTex->getCreationParameters().format);
+        //fill page table and pack present mips into physical addr texture
+        for (uint32_t i = 0u; i < levelCount; ++i)
         {
-            const uint32_t w = std::max(extent.width>>(_subres.baseMipLevel+i),1u);
-            const uint32_t h = std::max(extent.height>>(_subres.baseMipLevel+i),1u);
-            const uint32_t pgSqrSz = core::roundUpToPoT((std::max(w,h)+m_pgSzxy-1u)/m_pgSzxy);
+            const uint32_t w = (std::max(extent.width>>(_subres.baseMipLevel+i),1u) + m_pgSzxy-1u) / m_pgSzxy;
+            const uint32_t h = (std::max(extent.height>>(_subres.baseMipLevel+i),1u) + m_pgSzxy-1u) / m_pgSzxy;
 
             uint32_t* const pgTab = reinterpret_cast<uint32_t*>(
                 reinterpret_cast<uint8_t*>(m_pageTable->getBuffer()->getPointer()) + m_pageTable->getRegions().begin()[i].bufferOffset
                 );
-            const uint32_t pitch = m_pageTable->getRegions().begin()[i].bufferRowLength;
+            const uint32_t pgtPitch = m_pageTable->getRegions().begin()[i].bufferRowLength;
             const uint32_t pgtH = m_pageTable->getRegions().begin()[i].imageExtent.height;
-            const uint32_t offset = (pitch*pgtH)*(*offsets)[i].layer() + (*offsets)[i].y()*pitch + (*offsets)[i].x();
-            for (uint32_t x = 0u; x < pgSqrSz; ++x)
-                for (uint32_t y = 0u; y < pgSqrSz; ++y)
-                    pgTab[offset + y*pitch + x] = (*offsets)[i].addr;
+            const uint32_t offset = (pgtPitch*pgtH)*pgtOffset.layer() + (pgtOffset.y()>>i)*pgtPitch + (pgtOffset.x()>>i);
+
+            for (uint32_t y = 0u; y < h; ++y)
+                for (uint32_t x = 0u; x < w; ++x)
+                {
+                    const uint32_t addr = 0u;//TODO alloc with pool alctr
+                    if (i==(levelCount-1u) && levelCount>_subres.levelCount)
+                    {
+                        //TODO alloc another address and store on upper 16 bits of `addr` (for mip-tail)
+                    }
+                    pgTab[offset + y*pgtPitch + x] = addr;
+
+                    core::vector3du32_SIMD physPg = pageCoords(addr, m_pgSzxy);
+                    for (const auto& reg : _img->getRegions())
+                    {
+                        if (reg.imageSubresource.mipLevel != (_subres.baseMipLevel+i))
+                            continue;
+
+                        auto src_txOffset = core::vector2du32_SIMD(x,y)*m_pgSzxy;
```

Comment:
arent you supposed to compute that while taking the padding into account?

like in `physAddrTexLayerSz` or `pageCoords` ?

Reply:
nah, it's `src_` (i.e. into image being packed) offset, there's no padding

Reply:
ah ok, didn't notice it was `src`